Add the concept of "advanced" commands.

riannucci opened the pull request:
This allows a program to define many subcommands, but have them omitted from the default `help` view, to keep the high-level subcommands directly visible to the user.

vadimsht commented on subcommands.go:
```diff
@@ -283,23 +299,29 @@ func wrapWithLines(s string) string {
+	LongDesc:  "Prints an overview of every commands or information about a specific command.\nPass -advanced to see help for advanced commands.",
+	CommandRun: func() CommandRun {
+		ret := &helpRun{}
+		ret.advanced = ret.Flags.Bool("advanced", false, "show advanced commands")
```
Would it be difficult to not register this flag if Application has no advanced commands defined?

maruel commented on subcommands.go:
```diff
@@ -118,36 +118,44 @@ func (c *Command) Name() string {
+//
+// TODO(maruel): Use termbox-go to enable coloring!
+func usage(out io.Writer, a Application, includeAdvanced bool) {
+	usageTemplate := `{{.Title}}
```
It's a long constant to put inside a function, that's the reason I had used a named variable.

riannucci replied:
I'm not sure what you mean? It was already inside the function. The first commit in this PR moved it out, but then I realized I could make the overall diff smaller. I can squash the two changes if you like.

maruel replied:
Ohh I only looked at one commit by error.

maruel commented on subcommands.go:
```diff
@@ -283,23 +307,29 @@ func wrapWithLines(s string) string {
-	ShortDesc:  "prints help about a command",
-	LongDesc:   "Prints an overview of every commands or information about a specific command.",
-	CommandRun: func() CommandRun { return &helpRun{} },
+	UsageLine: "help <command> [-advanced]",
```
"help [-advanced] <command>"

maruel commented on subcommands.go:
```diff
@@ -283,23 +307,29 @@ func wrapWithLines(s string) string {
+	LongDesc:  "Prints an overview of every command or information about a specific command.\nPass -advanced to see help for advanced commands.",
+	CommandRun: func() CommandRun {
+		ret := &helpRun{}
+		ret.advanced = ret.Flags.Bool("advanced", false, "show advanced commands")
```
ret.Flags.BoolVar(&ret.advanced, ...)  ?

riannucci replied:
_shrug_ was following the rest of the file's style

maruel commented on subcommands.go:
```diff
@@ -199,9 +223,9 @@ func FindNearestCommand(a Application, name string) *Command {
 
 	// Calculate the levenshtein distance and take the closest one.
-	var closestD int = 1000
+	var closestD = 1000
```
closestD := 1000

I'm not sure why I had written the code the way I did previously.

maruel commented on subcommands.go:
```diff
@@ -120,18 +121,41 @@ func (c *Command) Name() string {
+	usageTemplate := `{{.Title}}
 
-Usage:  {{.GetName}} [command] [arguments]
+Usage:  {{.Name}} [command] [arguments]
```
why this change?

riannucci replied:
Because I'm no longer passing the application object. Doesn't make sense to have the map keys named 'GetBlah'

riannucci commented:
Fixed, but I'll squash when it lgty.

riannucci commented:
Y/N/Maybe?

riannucci commented:
Squashed to make change clearer

maruel commented on subcommands.go:
```diff
@@ -120,18 +121,41 @@ func (c *Command) Name() string {
-	tmpl(out, usageTemplate, a)
+
+	allCmds := a.GetCommands()
+	cmds := []*Command(nil)
```
I prefer for uninitialize slices:
`var cmd []*Command`

(this is the only case where I prefer the var form)

maruel reviewed (approved):
lgtm with one style nit.

riannucci committed:
Add the concept of "advanced" commands.

This allows a program to define many subcommands, but have them
omitted from the default help view, to keep the high-level
subcommands directly visible to the user.

riannucci commented:
Nit fixed!
